apps/aecore/src/aec_conductor.erl

```diff
@@ -889,31 +906,22 @@ flush_candidate() ->
         ok
     end.
 
-%% NG-TODO: This should be cleaned up
-setup_loop(State, IsLeader, block_created) ->
-    State1 = State#state{consensus = #consensus{leader = IsLeader}},
-    case IsLeader of
-        true ->
-            aec_block_generator:start_generation(),
-            State2 = start_micro_signing(State1),
-            start_mining(State2);
-        false ->
-            start_mining(State1)
-    end;
-setup_loop(State, IsLeader, block_received) ->
-    State1 = State#state{consensus = #consensus{leader = IsLeader}},
-    State2 = start_mining(State1),
-    case IsLeader of
-        true  -> ok;
-        false -> aec_block_generator:stop_generation()
-    end,
-    State2;
-setup_loop(State, IsLeader, micro_block_created) ->
-    State1 = State#state{consensus = #consensus{leader = IsLeader}},
-    State2 = start_mining(State1),
-    start_micro_sleep(State2);
-setup_loop(State, IsLeader, micro_block_received) ->
-    start_mining(State#state{consensus = #consensus{leader = IsLeader}});
-setup_loop(State, IsLeader, block_synced) ->
-    start_mining(State#state{consensus = #consensus{leader = IsLeader}}).
-
+setup_loop(State, RestartMining, IsLeader, Origin) ->
+    State1 = State#state{ consensus = #consensus{ leader = IsLeader } },
+    State2 =
+        case Origin of
+            block_created when IsLeader ->
+                aec_block_generator:start_generation(),
+                start_micro_signing(State1);
+            block_received when not IsLeader ->
+                aec_block_generator:stop_generation(),
+                State1;
+            micro_block_created when IsLeader ->
+                start_micro_sleep(State1);
+            _Other ->
```

Comment:
I wonder if it were possible to avoid the open ending case clause i.e. explicitly mention all potential values, as this would avoid typos or refactorings to change semantics of this code.